crypto/src/main/java/org/cryptimeleon/incentive/crypto/IncentiveSystem.java

```diff
@@ -109,9 +123,22 @@ public JoinRequest generateJoinRequest(ProviderPublicKey pk, UserKeyPair ukp, Pr
         UserPublicKey upk = ukp.getPk();
         UserSecretKey usk = ukp.getSk();
 
+
+
         // generate random values needed for generation of fresh user token using PRF hashThenPRFtoZn, user secret key is hash input
         IssueJoinRandomness R = computeIssueJoinRandomness(ukp.getSk(), promotionParameters);
```

Comment:
Doesn't all this mean that every time I enter a store, I'll broadcast the same blindedUpk, genesis signature, and commitment? Privacy-wise, that's not the best idea 😉 
I think the join process should not be deterministic. Just generate the random exponents on-the-fly instead of through the PRF.

Reply:
Haha, valid point! :smile: 

I'll remove the pseudorandomness from the clients calls and instead output and cache the randomness

Reply:
I think i'll just keep the IssueJoinRandomness class, but make it truly random, output it on generateJoinRequest and input it to the handle Response class. My proposed refactoring of having an instance of some IssueJoinClient class could make this a little nicer in the future